src/opm/output/eclipse/LoadRestart.cpp

```diff
@@ -1403,6 +1404,62 @@ namespace {
         smry.update(key("GITH"), xgrp[VI::XGroup::index::HistGasInjTotal]);
     }
 
+
+    void restore_udq(::Opm::SummaryState&             smry,
+                     const ::Opm::Schedule&           schedule,
+                     const std::shared_ptr<RestartFileView>& rst_view)
+    {
+        if (!rst_view->hasKeyword<std::string>(std::string("ZUDN")))
+            return;
+
+        const auto& intehead = rst_view->getKeyword<int>("INTEHEAD");
+        const auto& zudn = rst_view->getKeyword<std::string>("ZUDN");
+        const auto& dudw = rst_view->getKeyword<double>("DUDW");
+        const auto& dudg = rst_view->getKeyword<double>("DUDG");
+        const auto& dudf = rst_view->getKeyword<double>("DUDF");
+
+        const auto sim_step = rst_view->simStep();
+        const auto& wnames = schedule.wellNames(sim_step);
+        const auto& groups = schedule.restart_groups(sim_step);
+        auto num_wells = intehead[VI::intehead::NWMAXZ];
+        auto num_groups = intehead[VI::intehead::NGMAXZ];
+
+        std::size_t udq_well_index  = 0;
+        std::size_t udq_group_index = 0;
+        std::size_t udq_field_index = 0;
+
+        for (const auto& udq : zudn) {
+            if (udq[0] == 'W') {
+                std::size_t dudw_offset = udq_well_index * num_wells;
+                for (std::size_t well_index = 0; well_index < wnames.size(); well_index++) {
+                    const auto& value = dudw[dudw_offset + well_index];
+                    if (value != ::Opm::UDQ::restart_default)
+                        smry.update_well_var(wnames[well_index], udq, value);
+                }
```

Comment:
We've tried to avoid that kind of explicit and error prone index arithmetic here.  Is there no way you could wrap this in a type that returns appropriately sized data windows in the style of the other vectors instead?

Reply:
I'll look into using the `Window`class.

Reply:
OK - I have implemented a `UDQVectors` class - it is not exactly an epitome of clearness now either - but hopefully an improvement.

Reply:
> it is not exactly an epitome of clearness now either - but hopefully an improvement.

I do think it's an improvement over the original version.  The nested
```C++
for ()
    if ()
        for ()
            if ()
```
structure does limit readability and I might possibly split the inner loops out to separate helpers if I were writing this myself.